OpenRA.Game/FileFormats/Png.cs

```diff
@@ -206,5 +228,116 @@ static bool IsPaletted(byte bitDepth, PngColorType colorType)
 
 			throw new InvalidDataException("Unknown pixel format");
 		}
+
+		public byte[] Save()
+		{
+			var pixelFormat = Palette != null ? PixelFormat.Format8bppIndexed : PixelFormat.Format32bppArgb;
+
+			// Save to a memory stream that we can then parse to add the embedded data
+			using (var bitmapStream = new MemoryStream())
+			{
+				using (var bitmap = new Bitmap(Width, Height, pixelFormat))
+				{
+					if (Palette != null)
+					{
+						// Setting bitmap.Palette.Entries directly doesn't work
+						var bPal = bitmap.Palette;
+						for (var i = 0; i < 256; i++)
+							bPal.Entries[i] = Palette[i];
+
+						bitmap.Palette = bPal;
+						var bd = bitmap.LockBits(new Rectangle(0, 0, Width, Height), ImageLockMode.WriteOnly,
+							pixelFormat);
+						for (var i = 0; i < Height; i++)
+							Marshal.Copy(Data, i * Width, IntPtr.Add(bd.Scan0, i * bd.Stride), Width);
+
+						bitmap.UnlockBits(bd);
+					}
+					else
+					{
+						unsafe
+						{
+							var bd = bitmap.LockBits(new Rectangle(0, 0, Width, Height), ImageLockMode.WriteOnly, pixelFormat);
+							var colors = (int*)bd.Scan0;
+							for (var y = 0; y < Height; y++)
+							{
+								for (var x = 0; x < Width; x++)
+								{
+									var i = y * Width + x;
+
+									// Convert RGBA to ARGB
+									colors[i] = Color.FromArgb(Data[4 * i + 3], Data[4 * i + 0], Data[4 * i + 1],
+										Data[4 * i + 2]).ToArgb();
+								}
+							}
+
+							bitmap.UnlockBits(bd);
+						}
+					}
+
+					bitmap.Save(bitmapStream, ImageFormat.Png);
+				}
+
+				if (!EmbeddedData.Any())
+					return bitmapStream.ToArray();
+
+				// Add embedded metadata to the end of the image
+				bitmapStream.Position = 0;
+				var outputStream = new MemoryStream();
```

Comment:
Missing using()

Reply:
The `BinaryWriter` disposes the stream at the end of its using - one of the reasons I don't like about it, but I didn't see a strong reason to rewrite @IceReaper's initial use of it as this whole method needs to be replaced in the short term to remove `System.Bitmap`.